go/vt/vtgate/planbuilder/gen4_planner.go

```diff
@@ -93,7 +93,13 @@ func gen4SelectStmtPlanner(
 			return primitive, nil
 		}
 	}
-	return plan.Primitive(), nil
+
+	primitive := plan.Primitive()
+	if rb, ok := primitive.(*engine.Route); ok {
```

Comment:
I think a comment here as to why the previous behaviour is desirable for JOINs and others would be good.